src/cargo/ops/cargo_install.rs

```diff
@@ -530,22 +531,40 @@ impl<'cfg, 'a> InstallablePackage<'cfg, 'a> {
         // duplicate "Updating", but since `source` is taken by value, then it
         // wouldn't be available for `compile_ws`.
         let (pkg_set, resolve) = ops::resolve_ws(&self.ws)?;
-        let mut sources = pkg_set.sources_mut();
 
         // Checking the yanked status involves taking a look at the registry and
         // maybe updating files, so be sure to lock it here.
         let _lock = self.ws.config().acquire_package_cache_lock()?;
 
-        for pkg_id in resolve.iter() {
-            if let Some(source) = sources.get_mut(pkg_id.source_id()) {
-                if source.is_yanked(pkg_id)? {
-                    self.ws.config().shell().warn(format!(
-                        "package `{}` in Cargo.lock is yanked in registry `{}`, \
-                         consider running without --locked",
-                        pkg_id,
-                        pkg_id.source_id().display_registry_name()
-                    ))?;
+        let mut sources = pkg_set.sources_mut();
+        let mut pending: Vec<PackageId> = resolve.iter().collect();
+        let mut results = Vec::new();
+        for (_id, source) in sources.sources_mut() {
+            source.invalidate_cache();
+        }
+        while !pending.is_empty() {
+            pending.retain(|pkg_id| {
+                if let Some(source) = sources.get_mut(pkg_id.source_id()) {
+                    match source.is_yanked(*pkg_id) {
+                        Poll::Ready(result) => results.push((*pkg_id, result)),
+                        Poll::Pending => return true,
+                    }
                 }
+                false
+            });
+            for (_id, source) in sources.sources_mut() {
+                source.block_until_ready()?;
+            }
+        }
+
+        for (pkg_id, is_yanked) in results {
+            if is_yanked? {
+                self.ws.config().shell().warn(format!(
+                    "package `{}` in Cargo.lock is yanked in registry `{}`, \
+                     consider running without --locked",
+                    pkg_id,
+                    pkg_id.source_id().display_registry_name()
+                ))?;
```

Comment:
This is a pretty substantial amount of code to duplicate with the package command. Can it be pulled out somewhere?

Reply:
Do you have a suggestion where it could go? I didn't immediately see a good place for common code between install and package.

Reply:
I don't think the exact home is too important.  I think it would be fine to just re-export `ops::cargo_package::check_yanked` in `ops/mod.rs` and import it in `cargo_install.rs`.  I would probably just add another parameter that is the "hint" message on how to resolve it (which is different between the two).

Reply:
Opened #10835